Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,332 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Metrics API open source project
+//
+// Copyright (c) 2018-2020 Apple Inc. and the Swift Metrics API project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Swift Metrics API project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Dispatch
+
+#if os(Linux)
+import Glibc
+#endif
+
+extension MetricsSystem {
+    /// `bootstrapWithSystemMetrics` is an one-time configuration function which globally selects the desired metrics backend
+    /// implementation, and enables system level metrics. `bootstrapWithSystemMetrics` can be called at maximum once in any given program,
+    /// calling it more than once will lead to undefined behaviour, most likely a crash.
+    ///
+    /// - parameters:
+    ///     - factory: A factory that given an identifier produces instances of metrics handlers such as `CounterHandler`, `RecorderHandler` and `TimerHandler`.
+    ///     - config: Used to configure `SystemMetrics`.
+    public static func bootstrapWithSystemMetrics(_ factory: MetricsFactory, config: SystemMetrics.Configuration) {
+        let factory = SystemMetricsFactory(factory: factory, config: config)
+        self.bootstrap(factory)
+        factory.poll()
+    }
+
+    internal class SystemMetricsFactory: MetricsFactory {
+        fileprivate let queue = DispatchQueue(label: "com.apple.CoreMetrics.SystemMetricsHandler", qos: .background)
+        fileprivate let timeInterval: DispatchTimeInterval
+        fileprivate let dataProvider: SystemMetrics.DataProvider
+        fileprivate let labels: SystemMetrics.Labels
+        fileprivate var task: DispatchWorkItem?
+        internal let underlying: MetricsFactory
+
+        init(factory: MetricsFactory, config: SystemMetrics.Configuration) {
+            self.underlying = factory
+            self.timeInterval = config.interval
+            if let dataProvider = config.dataProvider {
+                self.dataProvider = dataProvider
+            } else {
+                #if os(Linux)
+                self.dataProvider = SystemMetrics.linuxSystemMetrics
+                #else
+                self.dataProvider = SystemMetrics.noopSystemMetrics
+                #endif
+            }
```

Comment:
feels to me like L46 -> L54 can move into the Configuration which can then have its DataProvider member be non-optional, i.e.

```swift
public struct Configuration {
  let interval: DispatchTimeInterval
  let dataProvider: SystemMetrics.DataProvider
  let labels: SystemMetrics.Labels

  public init(pollInterval interval: DispatchTimeInterval = .seconds(2), dataProvider: SystemMetrics.DataProvider? = nil, labels: Labels) {
    self.interval = interval
    if let dataProvider = config.dataProvider {
      self.dataProvider = dataProvider
    } else {
       #if os(Linux)
       self.dataProvider = SystemMetrics.linuxSystemMetrics
       #else
       self.dataProvider = SystemMetrics.noopSystemMetrics
       #endif
    }
    self.labels = labels
  }
}
```